### ISSUE TITLE
Switching branches should always Reset Local

### ISSUE THREAD

**Issue by pokrakam** [abapGit/abapGit#3684]
This is partly triggered by the interesting discussion on #3668 

A `git checkout myotherbranch` command will always make the local files / working tree match the target branch. abapGit's "Reset Local" will produce the equivalent behaviour, so I believe this should be applied after switching branches. 

Current behaviour is confusing. When switching branches, abapGit provides a variety of different options, particularly when files have been added/deleted. See test repo at https://github.com/pokrakam/abapGitTestBranch
Branch `different` only has a difference in the report
Branch `feature` has a difference and an additional file

When switching from `master` to `different` it offers to pull -> not correct but OK
when switching from `master` to `feature` it offers to pull or stage and indicates the feature object as "D"/delete
when switching from `feature` to `master` it offers to pull or stage and indicates the feature object as an addition

In all scenarios performing a "Reset Local" will do exactly what `git checkout branch` does. 
Opinions?

**Comment by mbtools** [comment]
imho switching branches, tag, or source (URL) should not update anything in my system. it should be a separate decision to change objects because I first need to see what might change (view diffs) and then decide what changes I want applied.

i do agree that having "pull" or "stage" available only under certain conditions makes things confusing (see https://github.com/larshp/abapGit/issues/3519). I rather have the options available all the time and we could remove "reset local" and "force stage".

**Comment by FreHu** [comment]
Agreed with removing secret functions like reset local and force stage. Maybe just have a warning message/popup when previously the situation would not be possible using pull/stage and allow to override.

**Comment by sbcgua** [comment]
Thoughts:
- Imho "reset local" is strange and and would be logical to always do it but ... there is one big difference. In AG is not a mere checkout, it is checkout + build !
- One of practical consequences: it will take more time to remove and re-deserialize object on quick switches. E.g. now, it is possible to "quick switch to a feature" without deleting other current object and then switch back and it takes less time. (though maybe the proper way would be to speedup deserialization ... instead of unusual commands)
- BTW: pull/stage are available always via hotkeys :) a bug-like feature :) So I guess displaying both of them if changes are available should be no big change.
- +1 for better messages before override

**Comment by pokrakam** [comment]
I’m looking at this from a regular git point of view, pretty much every other git client out there functions this way. All the git clients I've used (native git, Eclipse, VSCode, XCode, BBEdit, SAP WebUI) will switch my entire project environment to represent the current state of the commit that the branch is pointing to, totally transparent. Git is a snapshot-based VCS, a branch represents a point-in-time snapshot of the entire repo. Switching to a branch is like teleporting to that commit. 

The default behaviour of AG is that it helpfully offers to merge parts of your current branch’s state to the branch your’e switching to. This is wrong! I ran into it again while working on my PDTS branch: Discovered an unrelated bug, created a new branch from master to fix it and checked it out in AG. The first thing AG wanted to do is add my PDTS classes into my issue branch. Had I not been paying attention, this would have mixed my new feature with bugfix code. Other git clients won’t let that happen by accident. Performing a “Reset local” deleted my PDTS branch’s objects, which is exactly what a branch switch should have done. 

@mbtools - what you suggests is diff and merge, a completely different operation than switching branches. If we perform the equivalent of “Reset Local” after switching branches, the options remain the same (reset local asks for confirmation), but at least the default behaviour matches Git. The other thing that Git doesn’t allow you to do is switch branches when you have uncommitted (or at least unstashed) changes, so perhaps this check should be added to abapGit? 

Pushing or pulling between or after checkouts does not make sense in git. Perhaps this is desired in order to use abapGit in a similar manner to a file-based VCS, or to use abapGit as a way to manually merge branches? IMHO both of these are ungitlike behaviours. 

But it does raise the issue that the branch management & merging of abapGit is not something I’m entirely comfortable with, but that’s quite possibly just my understanding of how it’s implemented. Personally I do that outside of abapGit, mostly using the command line or Eclipse.

@sbcgua - good point about performance. Personally I don’t switch branches often in abapGit so this was never an issue. I just started playing around with AG branches for the first time (hence this issue). And I have noticed that the branch overview of abapGit is particularly intensive. I use Eclipse or the command line to do the ‘git’ stuff and mainly use abapGit to keep SAP in sync. I wonder if it’s feasible when switching branches to do a diff first and only retrieve/process the differing objects?

FWIW, on my command line a diff takes under 0.02 seconds. I can't easily see this in abapGit without switching branches:
![image](https://user-images.githubusercontent.com/25539927/89543493-ec078b00-d7f8-11ea-939d-b2ee90372744.png)

The branch overview takes over a minute for abapGit and doesn't show the graph on my system (too big I guess). So in the current setup switching branches is the only way to see this info and I guess it must stay in some form for now,  but I think this should move to the branch overview. This also relates to #3688

**Comment by larshp** [comment]
the regular git clients will change the entire project environment to represent the current state, when the operation does not cause loss of changed files (or at least thats my experience)

I'm okay with changing the pull/checkout behavior in abapGit under the same assumption.

**Comment by mbtools** [comment]
Let’s say we do the reset local, if there are no current diffs (ie uncommitted changes). If I see it right, there would be no need for popups since there are no changes to be overwritten. And due to the reset local, there would be no diffs afterwards. 

The question is then how do I see the diff between branches?

**Comment by habrahams** [comment]
Quite an interesting topic I also thought about the other day before I realised no branches are saved locally.

If AG does in future decide to do a reset local/synchronise objects on branch switch, I agree with @larshp that the switch should not cause loss of locally changed files. Also as @mbtools mentions, AG does not currently store copies of objects locally and thus a diff between branches is not possible.
This said, if this were to be implemented, then stashing of changes would also have to implemented, and a way to diff between current branch state and stashed changes. This implies that local copies would have to be created for the stashed changes. The next question would then be is, if local copies are kept for stashes, why not for branches as well? 
In the end if one was to go down this route you might end up breaking the first 3 of AG's design goals.

**Comment by pokrakam** [comment]
Interesting discussion, exactly what I was hoping for :) 
Rather than a wall of text, I'll add a couple of responses. This issue was intended as a discussion point, because implementing such a change does need several components, which should probably be raised as separate issues. 

For starters, I also agree with @larshp and @habrahams , that was precisely my suggestion when I wrote 
> Git doesn’t allow you to do is switch branches when you have uncommitted (or at least unstashed) changes, so perhaps this check should be added to abapGit?

I think this is important, switching branches should always do this regardless of whether we change the behaviour. Switching forks is a similar case, I think that should do the same, even though it does diverge from Git. Maybe that part of the discussion belongs to #3688

**Comment by larshp** [comment]
I suggest the following 3 things, to start out:

A: having "pull" or "stage" available only under certain conditions makes things confusing. I rather have the options available all the time and we could remove "reset local" and "force stage". (copy pasta'ed from @mbtools comment above)

B: have pull do reset local if needed, ie. also delete local objects at pull

C: create new issue regarding performance of branch overview

**Comment by pokrakam** [comment]
@mbtools, I agree and understand abapGit does not store a local repo. As I see it, the SAP system is the equivalent to the working tree and GitHub etc. is the Git database. See diagram at bottom of [this page](https://git-scm.com/book/en/v2/Getting-Started-What-is-Git), SAP+abapGit represents the working tree and the staging area, the database is on GitHub. I suppose it's technically possible to store the entire repo, which could also be done by merging some of abapGitServer's functionality into abapGit. But this is a big undertaking and I think the current model works OK for now.

To see the diff between branches, you have to go to the branch overview and merge one branch onto another, on the confirmation screen you get a side-by-side list of all changes. IMHO this could be better represented by the regular diff list we get when we open a repo and there are changes in the system. Also, the abapGit repo does not work very well on this page, better to try it with a smaller repo.

**Comment by pokrakam** [comment]
Just to make it a little more visual, here is an example from [abapGitTestBranch](https://github.com/pokrakam/abapGitTestBranch), which I have cloned onto my local machine. 

Branch `feature` contains a PROG `zbranchtest2` that is not in `master`. 
When I edit it and then checkout another branch that doesn't contain the file, here is how git responds:
![image](https://user-images.githubusercontent.com/25539927/89577744-0a857a80-d829-11ea-8b04-ba94ba765808.png)

If I then commit, and it lets me checkout `master`, upon which `zbranchtest2` vanishes.

Doing the same in abapGit however will happily switch, leaving my change uncommitted and instead offer to add it to my newly checked out branch:
![image](https://user-images.githubusercontent.com/25539927/89579296-70730180-d82b-11ea-8ec7-e45e4e42b2f2.png)

So if there is consensus to change abapGit behaviour to align closer with git, then there are a few issues that could be opened. I think it's important to warn when there are uncommitted changes before switching branches, regardless of whether we change switching behaviour.

**Comment by flaiker** [comment]
Another point for me is that the "working directory" is shared between multiple developers which is not the case in normal git. When multiple developers develop on a branch each in the same system you currently just need to switch to the correct branch and filter the (typically very large) staging screen by transport. In this scenario you never pull anything from the remote and just push. Unless the support for this (I imagine quite common?) scenario was to be dropped, _switching branches should never Reset Local_ in my opinion.

**Comment by pokrakam** [comment]
Isn't that an unusual / ungitlike use of branches? I would imagine multiple devs having a branch each on the same SAP system could become very difficult to manage. 

The scenarios I thus far envisaged for abapGit are: 

1. Each developer has their own SAP system, as is the case with open source on GitHub. Very gitlike.
2. Using Git to manage common components in a multi-system landscape. Also git-friendly and an excellent use of AG
3. Using Git to manage concurrent development when switching between versions, e.g. production fix while working on a new feature

In the third scenario, I would think that all developers in the same system MUST work on the same branch. But then I don't have much real world experience of this scenario with many devs.
I guess the way to keep things manageable is to have smaller repos, so that only small teams are impacted. If a production fix on the dunning report is needed then only the dunning repo needs to be branch-switched and a small team of devs can coordinate this among themselves. I'd imagine this is easier to do than hand-picking large branches with hundreds of objects.

**Comment by flaiker** [comment]
> Isn't that an unusual / ungitlike use of branches? 

It's very un-git-like.

> I would imagine multiple devs having a branch each on the same SAP system could become very difficult to manage.

If you have at maximum one branch per transport objects are exclusively locked in that transport and therefore in that branch (with exceptions where a transport only locks a subset of a file, for example LIMU-METH).

> 1. Each developer has their own SAP system, as is the case with open source on GitHub. Very gitlike.

This still seems impossible to me in an ERP context from multiple standpoints (infrastructure, processes like customizing synchronisation / master and application data replication, licensing, interfaces to external systems...).

> I guess the way to keep things manageable is to have smaller repos, so that only small teams are impacted. If a production fix on the dunning report is needed then only the dunning repo needs to be branch-switched and a small team of devs can coordinate this among themselves. I'd imagine this is easier to do than hand-picking large branches with hundreds of objects.

I would argue the opposite. If a repo is quite large it becomes increasingly improbable that two changes affect the same objects. But I also don't understand how you can do branch switching or pulling changes at all in this case. Switching your centralized dev environment to the state of production seems like madness to me. To do the hotfix you lock every reset object in a transport that you don't actually want to import anywhere. Objects which might or are very likely to be already locked in some other transport. Also what about the state of the QAS system in between? I imagine if you use the default 3-system CTS-based customer landscape you cannot really pull anywhere at all. Which results in abapGit just being a one-way attachment to the transport system to organize development.

Disclaimer: I don't actually have experience in working with multiple developers on the same abapGit repo either and they would likely just run away after seeing me mostly looking at the loading animation. It is the same scenario though for developing multiple isolated features at the same time in different branches/transports. I just try to get the benefits of git without changing the landscape which results in this compromise.

**Comment by pokrakam** [comment]
I have worked with such setups without Git quite, using a 2- or 3-development system landscape and it can get very messy if not done right, but it can work (and Solution Manager also supports several strategies).
With only a single dev system, it does become complicated. I would envisage that this could only be done via ToCs or Transport surgery. The process is something like: 
- Identify objects that need to be changed
- Create a repo with the affected package (`master` becomes your current development state).
- Create a new branch `fix`
- Import production objects.
- Do fix
- Merge fix branch into master.   <-- and exactly this step is where abapGit shines!

As far as CTS goes, one way would be to have a Transport per repo/package. Before switching branches one would need to delete the current TR, which would automatically result in the creation of a new one for the fix. Then once that has been transported, switch to `master`, thereby recreating a valid TR that should (hopefully) be equivalent to the old one.

Many ways to deal with that, and many ways to mess it up :-)

**Comment by FreHu** [comment]
What if we had two separate buttons in the branch switch dialog:

*   switch
*   switch and pull

**Comment by larshp** [comment]
@pokrakam I think pull is always an option now, does this fix the issue?

**Comment by larshp** [comment]
close?

**Comment by FreHu** [comment]
sorry, wrong issue
